### PR TITLE
Create brainpass apis

### PR DESCRIPTION
**Pull request by AshakaE**
# BrainPass API

_Used to store and track brainpass metadata, and also retrieve mint data for brainpass contract_

_How this works, we listen for NFT blockchain events through Alchemy, and store the NFT metadata on IPFS and also store the return hash in our db to keep track such that when external clients(OpenSea) use the base url with a specific nft id `3.json`, we do the search for the hash in our api and serve the json response from pinata_

## How should this be tested?

1. `REST Api` to be used in the contract
```bash
curl --location 'http://apiUrl/12345.json'
```
default response
```json
{
    "description": "Edit Wikis For Life",
    "image": "https://ipfs.io/ipfs/QmQqHUeQJXVA2x88Se7g1ZcbtEvWKRiMbhdSE6FggMK1Kf",
    "name": "GeneralBrainPass-1",
    "attributes": [
        {
            "display_type": "boost_percentage",
            "trait_type": "Subscription Days",
            "value": 30
        }
    ]
}
```

2. `Query` to track user mints in dashboard
```gql
{
  retrieveBrainPass(address: "string") {
    transactionHash
  }
}
```

3. _Endpoint for alchemy webhook_ - `BASE_URL/brainpass/events`

## Notes or observations

_This is a working idea, still a lot of refactoring an unit tests to be done_

## Linked issues

closes https://github.com/EveripediaNetwork/issues/issues/1447

### CODE REVIEW

**Comment by kesar** [comment]
i think clients expect 1.json and so on not just /1 

also worth mentioning that this is not the proper solution. Proper solution should be pushing those 1.json file into ipfs / pinata and using ipfs in the contract.

it might be fine as temporal solution, but after this, start working in the ipfs version. 

https://moralis.io/ipfs-nft-how-to-use-ipfs-for-nft-metadata/
https://docs.ipfs.tech/how-to/best-practices-for-nft-data/
https://medium.com/scrappy-squirrels/tutorial-nft-metadata-ipfs-and-pinata-9ab1948669a3

etcetc

File: src/App/BrainPass/brainPass.resolver.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+  @Query(() => [BrainPass], { name: 'retrieveBrainPass' })
+  async retrieveBrainPass(
+    @Args('address', { type: () => String }) address: string,
+  ) {
+    return this.brainPassRepository.getBrainPassByAddress(address)
+  }
+
+  @Mutation(() => BrainPass, { name: 'storeBrainPass' })
+  async storeBrainPass(
```
Comment (kesar) [inline src/App/BrainPass/brainPass.resolver.ts:18]: what's the security for this?

Comment (AshakaE) [reply src/App/BrainPass/brainPass.resolver.ts:18]: No added security yet, I'm following the expectations from @OleanjiKingCode and @Emmanueldmlr , subsequent updates will be added

File: src/App/BrainPass/brainPass.repository.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import { Controller } from '@nestjs/common'
+import { DataSource, Repository } from 'typeorm'
+import BrainPass from '../../Database/Entities/brainPass.entity'
+import BrainPassDto from './brainPass.dto'
+
+@Controller('relayer')
```
[the remark this answers is not in the document]
Comment (AshakaE) [reply src/App/BrainPass/brainPass.repository.ts]: 🤦

**Comment by kesar** [comment]
also ive been thinking this a bit more and the mutation wont work. anyone can just mint the nft using blockchain and this endpoint will never be hitted. 

service needs to listen for the particular event of minting and generate the .json file (or simulate it using the api)

**Comment by AshakaE** [comment]
> also ive been thinking this a bit more and the mutation wont work. anyone can just mint the nft using blockchain and this endpoint will never be hitted.
> 
> service needs to listen for the particular event of minting and generate the .json file (or simulate it using the api)

Perhaps https://web3js.readthedocs.io/en/v1.10.0/web3-eth-contract.html?highlight=transfer%20event#events ?

**Comment by kesar** [comment]
i guess something similar to how we wait for wiki posts, but for nft mints

**Comment by Emmanueldmlr** [comment]
> also ive been thinking this a bit more and the mutation wont work. anyone can just mint the nft using blockchain and this endpoint will never be hitted.
> 
> service needs to listen for the particular event of minting and generate the .json file (or simulate it using the api)

Great point.

**Pull request merged**